### ISSUE TITLE
Cannot build libtorch: SLEEF does not allow in-source builds

### ISSUE THREAD

**Issue by HerculesJack** [pytorch/pytorch#17181]
## 🐛 Bug

See below: `pytorch` was built successfully, but when I tried to build `libtorch`, it says `Cannot build libtorch: SLEEF does not allow in-source builds`.

## To Reproduce

```
conda create -n temp python=3.6
source activate temp
conda install numpy pyyaml mkl mkl-include setuptools cmake cffi typing
git clone --recursive https://github.com/pytorch/pytorch
export NO_CUDA=1
cd pytorch/
python setup.py install
python -m tools.build_libtorch
```

## Expected behavior

```
-- Found Numa  (include: /usr/include, library: /usr/lib64/libnuma.so)
-- Using third party subdirectory Eigen.
-- Could NOT find pybind11 (missing: pybind11_DIR)
-- System pybind11 found
-- pybind11 include dirs: /global/homes/h/hejia/pytorch.2/pytorch/torch/include
-- Adding OpenMP CXX_FLAGS: -fopenmp
-- Will link against OpenMP libraries: /usr/lib64/gcc/x86_64-suse-linux/4.8/libgomp.so;/usr/lib64/libpthread.so
-- 
-- ******** Summary ********
--   CMake version         : 3.12.2
--   CMake command         : /global/u2/h/hejia/.conda/envs/temp/bin/cmake
--   System                : Linux
--   C++ compiler          : /usr/bin/c++
--   C++ compiler version  : 4.8.5
--   CXX flags             :   -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -Wnon-virtual-dtor
--   Build type            : Release
--   Compile definitions   : TH_BLAS_MKL
--   CMAKE_PREFIX_PATH     : /usr/common/software/python/2.7-anaconda-4.4/bin/../
--   CMAKE_INSTALL_PREFIX  : /global/homes/h/hejia/pytorch.2/pytorch/torch/build_libtorch
--   CMAKE_MODULE_PATH     : /global/homes/h/hejia/pytorch.2/pytorch/cmake/Modules
-- 
--   ONNX version          : 1.4.1
--   ONNX NAMESPACE        : onnx_torch
--   ONNX_BUILD_TESTS      : OFF
--   ONNX_BUILD_BENCHMARKS : OFF
--   ONNX_USE_LITE_PROTO   : OFF
--   ONNXIFI_DUMMY_BACKEND : OFF
-- 
--   Protobuf compiler     : 
--   Protobuf includes     : 
--   Protobuf libraries    : 
--   BUILD_ONNX_PYTHON     : OFF
-- Could not find CUDA with FP16 support, compiling without torch.CudaHalfTensor
-- Removing -DNDEBUG from compile flags
-- MAGMA not found. Compiling without MAGMA support
-- Could not find hardware support for NEON on this machine.
-- No OMAP3 processor on this machine.
-- No OMAP4 processor on this machine.
-- AVX compiler support found
-- AVX2 compiler support found
-- Atomics: using GCC intrinsics
-- Found a library with BLAS API (mkl).
-- Found a library with LAPACK API (mkl).
disabling CUDA because NOT USE_CUDA is set
-- CuDNN not found. Compiling without CuDNN support
disabling ROCM because NOT USE_ROCM is set
-- MIOpen not found. Compiling without MIOpen support
-- OpenMP lib: provided by compiler
-- VTune profiling environment is unset
-- Found MKL-DNN: TRUE
-- GCC 4.8.5: Adding gcc and gcc_s libs to link line
-- NUMA paths:
-- /usr/include
-- /usr/lib64/libnuma.so
disabling CUDA because USE_CUDA is set false
CMake Error at third_party/sleef/CMakeLists.txt:27 (message):
  SLEEF does not allow in-source builds.

  You can refer to doc/build-with-cmake.md for instructions on how provide a
  separate build directory.  Note: Please remove autogenerated file
  `CMakeCache.txt` and directory `CMakeFiles` in the current directory.
```

## Environment

```
PyTorch version: 1.1.0a0+70ee257
Is debug build: No
CUDA used to build PyTorch: None

OS: SUSE Linux Enterprise Server 12 SP3
GCC version: (SUSE Linux) 4.8.5
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.1.0a0+70ee257
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] torch                     1.1.0a0+70ee257           <pip>
```

## Additional context

See #17139. I encountered the same issue with an older version of `pytorch`. There I solved it with `tools/build_pytorch_libs.sh`, but this file has been removed in the latest version, and I don't know how to let `tools/build_libtorch.py` use a separate build directory. Could anyone give me some ideas on this? Thanks in advance!

**Comment by soumith** [comment]
please follow along here as to how we build libtorch: https://github.com/pytorch/builder/blob/master/manywheel/build_common.sh#L134

**Comment by nicolone** [comment]
Sleef doesn't support nesting. I could fix it similarly to https://ceres-solver-review.googlesource.com/c/ceres-solver/+/9780:

- Using sleef_[SOURCE/BINARY]_DIR (which are defined by CMake when
  project(sleef) is called, in favour of CMAKE_[SOURCE/BINARY]_DIR
  enables sleef to be nested within (and built by) a larger CMake
  project (which also contains other projects).
- CMAKE_[SOURCE/BINARY]_DIR always refers to the top-level source
  and binary directories (i.e. the first encountered), as a result if
  sleef is a nested project within a larger project, these would not
  correctly identify the source/binary directories for sleef (as they
  would refer to the root project in which sleef is nested).
- Using sleef_[SOURCE/BINARY]_DIR should ensure that sleef always uses
  the correct source/binary directories, irrespective of whether sleef
  is nested or not.

```

[sleef_patch.txt](https://github.com/pytorch/pytorch/files/2979917/sleef_patch.txt)

**Comment by ezyang** [comment]
cc @nwsterling @AshStuff @tastyminerals

**Comment by ezyang** [comment]
@nicolone That patch looks reasonable to me. Would you mind opening a PR for it?

**Comment by rgommers** [comment]
@ezyang this is still a valid SLEEF bug (and there is an issue for it: https://github.com/shibatch/sleef/issues/249), but for PyTorch the problem got fixed 10 days ago by commit 794ee6d00c. Do you ant to keep this open?

**Comment by ezyang** [comment]
Nope, this seems like a good enough resolution to me. It will be cool if sleef fixes it but that probably isn't a high priority issue for us anymore.
